MBS-12761 / MBS-14180: Convert recording edit form to React

reosarevok opened the pull request:
### Implement MBS-12761, MBS-14180

# Description
Should be mostly a 1:1 conversion, except that this also blocks submission at the JS level for more invalid data issues (empty artist credit, invalid length or ISRC), which we didn't do in the TT form.

A commit implements MBS-14180 to drop classical guess feat - this simplifies the rest of the `guessFeat` update somewhat.

# Testing
Tested adding and editing recordings, including:
* Basic add / edit with just title and artist
* Using the guess feat button to move featured artists from the title to the AC, both when they are linked as a relationship (artist entity is preselected) and when they are not (only artist name is filled)
* Not being able to submit the form without selecting an AC at all (in TT it is allowed and reloads with an error)
* Adding, editing and deleting ISRCs, including validation and rejection of incorrect ISRCs
* Adding, editing and deleting track lengths, including validation and rejection of incorrect lengths
* Edit note errors are shown, including when opening the page (for `/create`) and when seeding.
* French translation still shows fine for the form elements after generating new `pot` files with `./po/update_pot.sh` and updating the `po` file with `msgmerge --no-wrap --no-fuzzy-matching --update po/mb_server.fr.po po/mb_server.pot`, then recompiling resources.

Additionally, the automatic tests we have for this still pass.

reosarevok committed:
Drop unneeded edit_form in event Role::Create

reosarevok committed:
Drop unneeded edit_form in event Role::Create

reosarevok committed:
MBS-12761: Convert recording edit form to React

This uses the new guessFeat introduced in the previous commit.
This is an initial conversion; further commits introduce extra
features like a lot more validation at the JS level before the
user submits the form.

reosarevok committed:
MBS-12761: Convert recording edit form to React

This uses the new guessFeat introduced in the previous commit.
This is an initial conversion; further commits introduce extra
features like a lot more validation at the JS level before the
user submits the form.

reosarevok committed:
Drop seemingly useless class

This class being added to the non-hydrated components is actually
confusing the hydrate utility and seems to have no actual use - we
never call this selector anywhere AFAICT.

reosarevok committed:
Drop seemingly useless class

This class being added to the non-hydrated components is actually
confusing the hydrate utility and seems to have no actual use - we
never call this selector anywhere AFAICT.

reosarevok committed:
Use useEffect instead of manifest to load form tools

reosarevok committed:
Use useEffect instead of manifest to load form tools

reosarevok committed:
Move formUnload code to hook

reosarevok committed:
Move formUnload code to hook

reosarevok committed:
Use React validation for name and AC

This removes the need for initializeValidation() and
replaces it with a more proper React way of disabling submit
using state.
I made updateNameFieldErrors reusable so we can update the
form when creating the state, otherwise the submit button
won't get disabled until the user writes something
in the name field.

reosarevok committed:
Use React validation for name and AC

This removes the need for initializeValidation() and
replaces it with a more proper React way of disabling submit
using state.
I made updateNameFieldErrors reusable so we can update the
form when creating the state, otherwise the submit button
won't get disabled until the user writes something
in the name field.

reosarevok committed:
Use form artist credit for ArtistCreditEditor

This avoids blanking the artist credit when reloading the page
on the recording creation page (now it will keep any AC the
user had entered).

reosarevok committed:
Use form artist credit for ArtistCreditEditor

This avoids blanking the artist credit when reloading the page
on the recording creation page (now it will keep any AC the
user had entered).

reosarevok committed:
Check for valid edit notes on RecordingEditForm

reosarevok committed:
Check for valid edit notes on RecordingEditForm

reosarevok commented:
I think this is ready for review now, although before merging the commits probably could be put together in a more reasonable way (not sure exactly whether to combine most into one or not)

mwiencek committed:
Ignore invalid artist IDs in `createInitialNamesState`

This can be triggered via
`/recording/create?edit-recording.artist_credit.names.0.artist.id=foo`.

mwiencek committed:
Ignore invalid artist IDs in `createInitialNamesState`

This can be triggered via
`/recording/create?edit-recording.artist_credit.names.0.artist.id=foo`.

mwiencek commented:
A few issues I found while testing:

* Seeding an invalid artist ID like `/recording/create?edit-recording.artist_credit.names.0.artist.id=foo` breaks the page (pushed a commit for this issue).
* If you submit an invalid ISRC, it doesn't show any errors on the page after the form reloads.
* If you submit an invalid length, it shows an error, but the submit button isn't re-enabled after you correct it.

reosarevok committed:
Move unformatTrackLength to its own file

I'll next use it as part of a new isInvalidLength JS check.

reosarevok committed:
Move unformatTrackLength to its own file

I'll next use it as part of a new isInvalidLength JS check.

reosarevok committed:
Make unformatTrackLength work consistently with Perl

Our unformatTrackLength JS implementation was a lot
different from the Perl one. This worked fine when it was
only used for the release editor, but now we want to also use it
to test for validity at the JS level, so it should probably
consider as valid the same things the Perl backend will.

This commit mostly changes the unformatTrackLength JS version
to work more like the Perl one. There's two things that the Perl
version did not allow that probably make sense, so I've
added them there: :SS times without minutes,
and just parsing a number of seconds.

I'm removing the test for 3723494 seconds from fields.js
because that's actually a larger number in ms than we accept
in the DB (and was being rejected by the new method).

Similarly, :10 should now be an accepted track time, so changing
that test to use a time that is actually still disallowed.

reosarevok committed:
Make unformatTrackLength work consistently with Perl

Our unformatTrackLength JS implementation was a lot
different from the Perl one. This worked fine when it was
only used for the release editor, but now we want to also use it
to test for validity at the JS level, so it should probably
consider as valid the same things the Perl backend will.

This commit mostly changes the unformatTrackLength JS version
to work more like the Perl one. There's two things that the Perl
version did not allow that probably make sense, so I've
added them there: :SS times without minutes,
and just parsing a number of seconds.

I'm removing the test for 3723494 seconds from fields.js
because that's actually a larger number in ms than we accept
in the DB (and was being rejected by the new method).

Similarly, :10 should now be an accepted track time, so changing
that test to use a time that is actually still disallowed.

reosarevok committed:
Check for valid recording length on RecordingEditForm

reosarevok committed:
MBS-12761: Move ISRC validation and changing to React state

This modifies FormRowTextList to make it properly usable inside
a larger form; rather than have its own state, it accepts
dispatch and state props. Additionally, I'm moving currentTextValues
outside the state since it is a constant, and just passing it as
a prop. This was in the state for ease of use in the reducer,
but it can just be passed by the dispatch as needed.

For uses where the FormRow still needs its own state (because
it's just a standalone component used in a TT form) I added
StandaloneFormRowTextlist, to which the old local state has moved.

I added isValidIsrc for JS-level ISRC validation; this combines
is_valid_isrc and format_isrc from Server::Validation since we don't
want to reject ISRCs that are valid once formatted by the system.
I expect it'd be more confusing for the user to automatically change
the format of the ISRC on the form itself so I'm letting it happen
at the Perl level as earlier.

reosarevok committed:
Also validate edit note on initial state

For creating standalone recordings, it makes sense to display
an error saying a note is required from the get go.
This additionally set errors for seeded bad edit notes.

reosarevok committed:
Also validate edit note on initial state

For creating standalone recordings, it makes sense to display
an error saying a note is required from the get go.
This additionally set errors for seeded bad edit notes.

mwiencek commented on root/static/scripts/recording/components/RecordingEditForm.js:
```diff
@@ -0,0 +1,496 @@
+      });
+      runNameReducer(nameStateCtx, action);
+
+      const nameState = nameStateCtx.read();
```
Prefer `nameStateCtx.final();` if you intend to not modify the value anymore.

reosarevok replied:
Seems sensible, I changed it for all four forms that did the same in a separate commit.

mwiencek commented on root/static/scripts/recording/components/RecordingEditForm.js:
```diff
@@ -0,0 +1,496 @@
+            </p>
+          </Bubble>
+        ) : null}
+        {/* TODO: Make this show */}
```
Is this TODO still a WIP?

mwiencek reviewed (commented):
A few other minor things I found...
 * If I enter an invalid ISRC, the margin or padding of the error message appears misaligned (it's in the center of the field, but should be left-aligned). Also, the server uses "This is not a valid ISRC.", while the JS uses "Invalid value."
 * Causing the bubble doc to change (by switching focus on another field) makes the bubble flicker and appear/disappear in the top left of the page.
 * If I seed an invalid length, it doesn't show an error message.

reosarevok commented on root/static/scripts/common/utility/unformatTrackLength.js:
```diff
@@ -7,25 +7,63 @@
+  }
 
-  return (hours + minutes + seconds) * 1000;
+  return false;
```
I can use `NaN` if you prefer, no problem with that.

reosarevok committed:
Show external links bubble on focus

reosarevok committed:
Show external links bubble on focus

reosarevok committed:
Use the three-argument form of `useReducer`

The previous code invokes `createInitialState` on every single render.
I noticed when changing the recording form that we still had
this one around.

reosarevok committed:
Use the three-argument form of `useReducer`

The previous code invokes `createInitialState` on every single render.
I noticed when changing the recording form that we still had
this one around.

reosarevok committed:
Ensure no accidental writes on nameStateCtx

We don't intend to change nameStateCtx after assigning it
to nameState in any of these forms, so it is safer to use final()
to make sure it's not writable anymore.

reosarevok committed:
Ensure no accidental writes on nameStateCtx

We don't intend to change nameStateCtx after assigning it
to nameState in any of these forms, so it is safer to use final()
to make sure it's not writable anymore.

reosarevok commented on root/static/scripts/recording/components/RecordingEditForm.js:
```diff
@@ -80,20 +81,40 @@ type StateT = {
+  const formCtx = mutate(form);
+  // $FlowExpectedError[incompatible-call]
+  const nameFieldCtx = formCtx.get('field', 'name');
+  updateNameFieldErrors(nameFieldCtx);
```
I guess we could leave the button for required fields. The same happens with the AC though which I don't think will be as obvious, so we should probably at most prevent sending and show pending errors in case of submission?

reosarevok committed:
Use useLayoutEffect for Bubble

useEffect causes an annoying flickering effect on changing
the bubble position that seems to be avoided by useLayoutEffect.

Thanks @mwiencek for the suggestion.

reosarevok committed:
Use useLayoutEffect for Bubble

useEffect causes an annoying flickering effect on changing
the bubble position that seems to be avoided by useLayoutEffect.

Thanks @mwiencek for the suggestion.

reosarevok committed:
Also validate length on initial state

This is useful only for seeded lengths, but it makes sense to
show an error for those when invalid.

reosarevok committed:
Also validate length on initial state

This is useful only for seeded lengths, but it makes sense to
show an error for those when invalid.

reosarevok committed:
Update guessFeat to use Flow and add non-ko version

We will need an updated guessFeat for the recording form on a
subsequent commit. Since this makes a fair amount of changes
and improves the tests, I'm splitting it into its own commit
for ease of review.

I'm not touching the existing knockout guessFeat, which will
still be needed in the release editor and the RG form
until those two get converted to React as well. Once that
is complete, we can remove the last two functions of the
guessFeat file and simplify the tests a bit.

reosarevok committed:
Update guessFeat to use Flow and add non-ko version

We will need an updated guessFeat for the recording form on a
subsequent commit. Since this makes a fair amount of changes
and improves the tests, I'm splitting it into its own commit
for ease of review.

I'm not touching the existing knockout guessFeat, which will
still be needed in the release editor and the RG form
until those two get converted to React as well. Once that
is complete, we can remove the last two functions of the
guessFeat file and simplify the tests a bit.

We cannot use flow strict since we still import jQuery here.

reosarevok committed:
MBS-14180: Drop classical featuring guess feature

This was very useful shortly after NGS migration, but it hasn't
been really needed for years since most of the old-style MB
classical data has been fixed. New classical releases (from
data outside MusicBrainz) do not use the old MB style this
is built to fix at all, so there's no need to keep it
and we can simplify the guess feat code.

reosarevok committed:
MBS-14180: Drop classical featuring guess feature

This was very useful shortly after NGS migration, but it hasn't
been really needed for years since most of the old-style MB
classical data has been fixed. New classical releases (from
data outside MusicBrainz) do not use the old MB style this
is built to fix at all, so there's no need to keep it
and we can simplify the guess feat code.

reosarevok committed:
Display doc bubble for AC field

This creates a reusable FormRowArtistCredit since we will want to
do the same for RGs and probably releases.

yvanzo reviewed (commented):
@reosarevok Besides conflicts to be resolved, the add/remove buttons and help bubble for the ISRCs field is  no longer translated.

yvanzo commented on root/static/scripts/edit/utility/isInvalidLength.js:
```diff
@@ -0,0 +1,15 @@
+import unformatTrackLength from '../../common/utility/unformatTrackLength.js';
+
+export default function isInvalidLength(length: string): boolean {
+  // If the unformatted length is false, then it's an unparseable date
```
Typo about _date_ versus _length_. If _unparseable_, is it _invalid_ then?

```suggestion
  // If the unformatted length is false, that’s because the length string is unparseable, hence invalid.
```

reosarevok replied:
Applied the change (well, as a multiline comment)

yvanzo commented on root/static/scripts/tests/utility/isInvalidLength.js:
```diff
@@ -0,0 +1,65 @@
+    'MM:SS is valid',
+  );
+  t.ok(
+    !isInvalidLength('00:57'),
```
The test description is about `:SS` on the line below and in the original file `root/static/scripts/tests/release-editor/utils.js`. The above test uses `00:23` to test `MM:SS` already.

```suggestion
    !isInvalidLength(':57'),
```

yvanzo commented on root/static/scripts/tests/utility/isInvalidLength.js:
```diff
@@ -0,0 +1,65 @@
+  );
+  t.ok(
+    !isInvalidLength('00:57'),
+    ':SS is valid',
```
This test is copied from the old `root/static/scripts/tests/release-editor/utils.js` or the new `root/static/scripts/tests/utility/unformatTrackLength.js`.
The reference to MBS-3352 should be copied too.

reosarevok replied:
I thought it was not relevant here and it was a historical reference only for a test in an old file, but if you think it matters, sure, will add it.

yvanzo commented on root/static/scripts/tests/utility/isInvalidLength.js:
```diff
@@ -0,0 +1,65 @@
+  );
+  t.ok(
+    !isInvalidLength('?:??'),
+    '?:?? is empty/unknown, but not invalid',
```
This test is copied from the old `root/static/scripts/tests/release-editor/utils.js` or the new `root/static/scripts/tests/utility/unformatTrackLength.js`.
The reference to MBS-5086 should be copied too.

yvanzo commented on root/static/scripts/tests/utility/isInvalidLength.js:
```diff
@@ -0,0 +1,65 @@
+    !isInvalidLength('00:57'),
+    ':SS is valid',
+  );
+  t.ok(
+    !isInvalidLength('85:23'),
+    'MM:SS is valid with more than 60 minutes',
+  );
+  t.ok(
+    !isInvalidLength('1:00:57'),
+    'HH:MM:SS is valid',
+  );
+  t.ok(
+    isInvalidLength('foo'),
+    'A random string is not valid',
+  );
+  t.ok(
+    isInvalidLength('10:80'),
+    'MM:SS is not valid with more than 60 seconds',
+  );
+  t.ok(
+    isInvalidLength(':80'),
+    ':SS is not valid with more than 60 seconds',
+  );
+  t.ok(
+    isInvalidLength('1:75:10'),
+    'HH:MM:SS is not valid with more than 60 minutes',
+  );
+  t.ok(
+    isInvalidLength('1000:35:10'),
+    'Times longer than the max number of ms we can store are invalid',
+  );
```
Among these new tests, those with valid results should be added to the tests for `unformatTrackLength` to make sure that its output does match the expected length.

I would suggest to add new tests through separate commits.

yvanzo commented on root/static/scripts/tests/utility/isInvalidLength.js:
```diff
@@ -0,0 +1,65 @@
+  );
+  t.ok(
+    isInvalidLength('1000:35:10'),
+    'Times longer than the max number of ms we can store are invalid',
```
Where/how is this limit defined? Please explain in code comment or commit message.

reosarevok replied:
It's just a check that allows only hours up to 3 digits - I think that's the reasoning, but I'm not 100% sure actually, so I made the test text about what I am sure of, and put the reasoning in the commit message.

yvanzo commented on root/static/scripts/edit/components/FormRowArtistCredit.js:
```diff
@@ -0,0 +1,40 @@
+import FieldErrors from './FieldErrors.js';
+import FormRow from './FormRow.js';
+
+component FormRowAreaAutocomplete(
```
Typo likely from copy-paste:

```suggestion
component FormRowArtistCredit(
```

yvanzo commented on root/static/scripts/edit/components/FormRowArtistCredit.js:
```diff
@@ -0,0 +1,40 @@
+  );
+}
+
+export default FormRowAreaAutocomplete;
```
Same typo as above:
```suggestion
export default FormRowArtistCredit;
```

yvanzo commented on root/static/scripts/edit/components/StandaloneFormRowTextList.js:
```diff
@@ -0,0 +1,61 @@
+/*
+ * @flow
```
Likely a stale copy-paste, any reason not to use `strict` here?

```suggestion
 * @flow strict
```

reosarevok replied:
No good reason, good catch!

yvanzo commented on root/static/scripts/edit/utility/guessFeat.js:
```diff
@@ -1,4 +1,5 @@
 /*
+ * @flow
```
Any reason not to use `strict`? Please explain in commit message.

reosarevok replied:
It's not compatible with the jQuery import, for example, since that is non-flow-typed (neither is `strict-local`). I can mention it in the message.

yvanzo commented on root/static/scripts/edit/components/FormRowTextList.js:
```diff
@@ -207,27 +207,20 @@ component FormRowTextList(
 }
 
-export component NonHydratedFormRowTextList(
+component FormRowTextList(
```
Please make the change from `NonHydratedFormRowTextList` to `StandaloneFormRowTextList` a separate commit.

Also, should `NonHydratedFormRowTextListSimple` be changed too? Worth a ticket?

reosarevok replied:
For the second part: I am doing that as part of the label one in https://github.com/metabrainz/musicbrainz-server/pull/3672 (https://github.com/metabrainz/musicbrainz-server/pull/3672/changes/9185aff3bc808c67b43c6187cba628fa878cf30a - but it will probably need some update on rebasing, plus fixing the missing Simple in the commit message...)

reosarevok replied:
Separating this is not particularly trivial - the change to add `StandaloneFormRowTextList` is quite connected with the change to `FormRowTextList` in general, and that with the `RecordingEditForm` changes. I guess I could split the `isValidIsrc` addition itself with some (non-trivial) effort, and I can for sure split the `isValidIsrc` test addition (did that now).

yvanzo reviewed (commented):
Some more comments about tests and code details.

(Still not finished with testing translation.)

reosarevok committed:
Check for valid recording length on RecordingEditForm

This adds a utility function isInvalidLength which just checks
whether unformatTrackLength is unable to parse the given length.

reosarevok committed:
Add tests for isInvalidLength

This adds tests for the new isInvalidLength function, and adds
a couple of the cases also to the tests for unformatTrackLength.

unformatTrackLength does not support more than 3 digits for the
hours in the length - my understanding is that this is a limit
to avoid trying to store values too large (in ms) for us to do so.

reosarevok committed:
Display doc bubble for AC field

This creates a reusable FormRowArtistCredit since we will want to
do the same for RGs and probably releases.

reosarevok committed:
MBS-12761: Move ISRC validation and changing to React state

This modifies FormRowTextList to make it properly usable inside
a larger form; rather than have its own state, it accepts
dispatch and state props. Additionally, I'm moving currentTextValues
outside the state since it is a constant, and just passing it as
a prop. This was in the state for ease of use in the reducer,
but it can just be passed by the dispatch as needed.

For uses where the FormRow still needs its own state (because
it's just a standalone component used in a TT form) I added
StandaloneFormRowTextlist, to which the old local state has moved.

I added isValidIsrc for JS-level ISRC validation; this combines
is_valid_isrc and format_isrc from Server::Validation since we don't
want to reject ISRCs that are valid once formatted by the system.
I expect it'd be more confusing for the user to automatically change
the format of the ISRC on the form itself so I'm letting it happen
at the Perl level as earlier.

reosarevok committed:
Add tests for isValidIsrc

yvanzo commented on root/static/scripts/tests/utility/isInvalidLength.js:
```diff
@@ -0,0 +1,69 @@
+  );
+  t.ok(
+    !isInvalidLength('15:23'),
+    'MM:SS is valid with more than 60 minutes',
```
Typo in copy-paste (as `15 < 60`):

```suggestion
    'MM:SS is valid',
```

reosarevok replied:
Fixed this, good catch. Cancelled the test workflow to save resources since this is the only change and tests passed before.

yvanzo reviewed (approved):
Thank you! It looks good now, just found one last typo. 🚢 

* I tested the recording editor manually.
* I didn’t test seeding which has been tested before.
* I used the fixed `xgettext-js` release 5.0.3 to test internationalization.

🏷️ I support moving on to the label edit form PR, because of the `ForRowTextListSimple` bit.
